File: python/aide_frame/args.py
"""
Common command-line argument handling for aide-frame applications.

Provides reusable argument definitions that apps can add to their parsers.

Usage:
    from aide_frame.args import add_common_args, apply_common_args

    parser = argparse.ArgumentParser(description='My App')
    add_common_args(parser)  # adds --log-level, --config
    parser.add_argument('--port', type=int, default=8080)  # app-specific
    args = parser.parse_args()
    apply_common_args(args)  # applies log level, loads config
"""

import argparse
from typing import Optional, List, Dict, Any

from .log import set_level
from .config import load_config


def add_common_args(parser: argparse.ArgumentParser,
                    include_log: bool = True,
                    include_config: bool = True,
                    config_default: str = 'config.json') -> None:
    """
    Add common aide-frame arguments to a parser.

    Args:
        parser: The ArgumentParser to add arguments to
        include_log: Add --log-level argument
        include_config: Add --config argument
        config_default: Default config file name
    """
    if include_log:
        parser.add_argument(
            '--log-level', '-l',
            default='INFO',
            choices=['DEBUG', 'INFO', 'WARNING', 'ERROR'],
            help='Log level (default: INFO)'
        )

    if include_config:
        parser.add_argument(
            '--config', '-c',
            default=config_default,
            help=f'Config file path (default: {config_default})'
        )


def apply_common_args(args: argparse.Namespace,
                      config_search_paths: Optional[List[str]] = None,
                      config_defaults: Optional[Dict[str, Any]] = None) -> Optional[Dict[str, Any]]:
    """
    Apply common aide-frame arguments.

    Args:
        args: Parsed arguments namespace
        config_search_paths: Additional paths to search for config file
        config_defaults: Default config values

    Returns:
        Loaded config dict if --config was used, None otherwise
    """
    # Apply log level
    if hasattr(args, 'log_level'):
        set_level(args.log_level)

    # Load config if argument exists
    config = None
    if hasattr(args, 'config'):
        search_paths = config_search_paths or []
        config = load_config(
            config_path=args.config,
            search_paths=search_paths,
            defaults=config_defaults or {}
        )

    return config

File: js/aide_frame/src/args.js
/**
 * Common command-line argument handling for aide-frame applications.
 *
 * Provides reusable argument definitions using the commander library.
 *
 * Usage:
 *   const { Command } = require('commander');
 *   const { args } = require('aide-frame');
 *
 *   const program = new Command();
 *   args.addCommonArgs(program);
 *   program.parse();
 *   const config = args.applyCommonArgs(program.opts());
 */

const path = require('path');
const { Command } = require('commander');
const { setLevel } = require('./log');
const { loadConfig } = require('./config');
const { ensureIcons } = require('./icon-generator');

/**
 * Add common aide-frame arguments to a commander program.
 *
 * @param {Command} program - The Commander program to add arguments to
 * @param {object} options - Options
 * @param {boolean} options.includeLog - Add --log-level argument (default: true)
 * @param {boolean} options.includeConfig - Add --config argument (default: true)
 * @param {boolean} options.includeIcons - Add --regenerate-icons argument (default: true)
 * @param {boolean} options.includePort - Add --port argument (default: true)
 * @param {string} options.configDefault - Default config file name (default: 'config.json')
 */
function addCommonArgs(program, options = {}) {
    const {
        includeLog = true,
        includeConfig = true,
        includeIcons = true,
        includePort = true,
        configDefault = 'config.json',
    } = options;

    if (includeLog) {
        program.option(
            '-l, --log-level <level>',
            'Log level (DEBUG, INFO, WARNING, ERROR)',
            'INFO'
        );
    }

    if (includeConfig) {
        program.option(
            '-c, --config <path>',
            'Config file path',
            configDefault
        );
    }

    if (includeIcons) {
        program.option(
            '--regenerate-icons',
            'Force regeneration of PWA icons'
        );
    }

    if (includePort) {
        program.option(
            '-p, --port <number>',
            'Override server port from config',
            parseInt
        );
    }
}

/**
 * Apply common aide-frame arguments.
 *
 * @param {object} opts - Parsed options from commander
 * @param {object} options - Additional options
 * @param {string[]} options.configSearchPaths - Additional paths to search for config file
 * @param {object} options.configDefaults - Default config values
 * @param {string} options.appDir - App directory for icon generation (optional)
 * @returns {object|null} Loaded config object if --config was used
 */
function applyCommonArgs(opts, options = {}) {
    const {
        configSearchPaths = [],
        configDefaults = {},
        appDir = null,
        systemDir = null,
    } = options;

    // Apply log level
    if (opts.logLevel) {
        setLevel(opts.logLevel);
    }

    // Load config if argument exists
    let config = null;
    if (opts.config) {
        config = loadConfig(opts.config, configDefaults, configSearchPaths);
    }

    // Generate icons if configured and appDir provided
    if (appDir && config && config.pwa) {
        let outputDir = null;
        const iconConfig = config.pwa.icon || {};

        if (systemDir && iconConfig.outputDir === 'system') {
            outputDir = path.join(systemDir, 'icons');
        }

        ensureIcons(appDir, config.pwa, opts.regenerateIcons || false, outputDir);
    }

    // Override port from CLI if specified
    if (opts.port && config) {
        config.port = opts.port;
    }

    return config;
}

/**
 * Parse arguments and apply common settings.
 * Convenience function that combines addCommonArgs and applyCommonArgs.
 *
 * @param {object} options - Options
 * @param {string} options.description - Program description
 * @param {string} options.version - Program version
 * @param {object} options.configDefaults - Default config values
 * @param {string[]} options.configSearchPaths - Config search paths
 * @returns {object} Object with { program, opts, config }
 */
function parseArgs(options = {}) {
    const {
        description = 'AIDE application',
        version = '0.1.0',
        configDefaults = {},
        configSearchPaths = [],
    } = options;

    const program = new Command();
    program
        .description(description)
        .version(version);

    addCommonArgs(program);
    program.parse();

    const opts = program.opts();
    const config = applyCommonArgs(opts, { configSearchPaths, configDefaults });

    return { program, opts, config };
}

module.exports = {
    addCommonArgs,
    applyCommonArgs,
    parseArgs,
};
